src/core_ocean/shared/mpas_ocn_wetting_drying.F

```diff
@@ -0,0 +1,710 @@
+
+      integer, intent(out) :: err !< Output: error flag
+
+      !-----------------------------------------------------------------
+      !
+      ! local variables
+      !
+      !-----------------------------------------------------------------
+
+      type (mpas_pool_type), pointer :: statePool, meshPool, tendPool
+      integer, dimension(:), pointer :: maxLevelCell
+      real (kind=RKIND), dimension(:), pointer :: sshSubcycleNew, bottomDepth
+      integer, pointer :: nCellsSolve
+      integer :: iCell, k
+      integer :: debugUnit
+      real (kind=RKIND), dimension(:,:), pointer :: layerThicknessCur, layerThicknessNew, layerThicknessTend
+      real (kind=RKIND) :: minThickness, layerThick
+      character (len=StrKIND) :: debugFilename
+      character (len=StrKIND), pointer :: config_time_integrator
+
+      !-----------------------------------------------------------------
+      !
+      ! call relevant routines for computing coefficients
+      !
+      !-----------------------------------------------------------------
+
+      call mpas_pool_get_subpool(block % structs, 'state', statePool)
+      call mpas_pool_get_subpool(block % structs, 'mesh', meshPool)
+      call mpas_pool_get_subpool(block % structs, 'tend', tendPool)
+
+      call mpas_pool_get_array(statePool, 'layerThickness', layerThicknessCur, timeLevel=1)
+      call mpas_pool_get_array(statePool, 'layerThickness', layerThicknessNew, timeLevel=2)
+      call mpas_pool_get_dimension(meshPool, 'nCellsSolve', nCellsSolve)
+      call mpas_pool_get_array(meshPool, 'maxLevelCell', maxLevelCell)
+      call mpas_pool_get_array(tendPool, 'layerThickness', layerThicknessTend)
+      call mpas_pool_get_array(statePool, 'sshSubcycle', sshSubcycleNew, 2)
+      call mpas_pool_get_array(meshPool, 'bottomDepth', bottomDepth)
+      call mpas_pool_get_config(ocnConfigs, 'config_time_integrator', config_time_integrator)
+
+      err = 0
+
+      call mpas_log_write( 'Verifying that cells are not dry...')
+
+      ! check to make sure that there is no layer that is too dry
+      minThickness = +1.0E34
+      do iCell = 1, nCellsSolve
+        do k = 1, maxLevelCell(iCell)
+          !call mpas_log_write( 'Layer thickness = $r', realArgs=(/ layerThick/))
+          ! use ssh as a proxy too for baroclinic mode
+          if (trim(config_time_integrator) == 'split_explicit') then
+            layerThick = min(layerThicknessNew(k, iCell), (sshSubcycleNew(iCell)+bottomDepth(iCell))/maxLevelCell(iCell))
+          else
+            layerThick = layerThicknessNew(k, iCell)
+          end if
+          minThickness = min(minThickness, layerThick)
+          if ( layerThick <  minHeight ) then
+            ! report error if layerThickness constraint is violated
+            if (err == 0) then
+              debugFilename = ocn_build_log_filename('mpas_ocean_layerThickness_block_stats_', block % blockID)
+              call mpas_new_unit(debugUnit)
```

Comment:
@xylar, I'm thinking this may be the offending code that causes the memory leak.

Reply:
... or other lines like ` call mpas_new_unit(...)`